man/man1/peval.1
.\"// $Id$
.\"// Copyright (c) 2001,2002                        RIPE NCC
.\"//
.\"// All Rights Reserved
.\"//
.\"// Permission to use, copy, modify, and distribute this software and its
.\"// documentation for any purpose and without fee is hereby granted,
.\"// provided that the above copyright notice appear in all copies and that
.\"// both that copyright notice and this permission notice appear in
.\"// supporting documentation, and that the name of the author not be
.\"// used in advertising or publicity pertaining to distribution of the
.\"// software without specific, written prior permission.
.\"//
.\"// THE AUTHOR DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING
.\"// ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS; IN NO EVENT SHALL
.\"// AUTHOR BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY
.\"// DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN
.\"// AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\"// OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"//
.\"// 
.\"//  Copyright (c) 1994 by the University of Southern California
.\"//  and/or the International Business Machines Corporation.
.\"//  All rights reserved.
.\"//
.\"//    Permission is hereby granted, free of charge, to any person obtaining a copy
.\"//    of this software and associated documentation files (the "Software"), to deal
.\"//    in the Software without restriction, including without limitation the rights
.\"//    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
.\"//    copies of the Software, and to permit persons to whom the Software is
.\"//    furnished to do so, subject to the following conditions:
.\"//
.\"//    The above copyright notice and this permission notice shall be included in
.\"//    all copies or substantial portions of the Software.
.\"//
.\"//    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
.\"//    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
.\"//    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
.\"//    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
.\"//    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
.\"//    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
.\"//    THE SOFTWARE.
.\"//
.\"//  Questions concerning this software should be directed to 
.\"//  irrtoolset@ripe.net
.\"//
.\"//  Author(s): Cengiz Alaettinoglu <cengiz@isi.edu>
.\"//             Katie Petrusha <katie@ripe.net>
.\"
.\"
.TH peval 1 local
.SH NAME
peval
.SH SYNOPSIS
.B peval
.RI [ options ]\ [<expression>]
.SH DESCRIPTION
.PP
peval is a low level policy evaluation tool 
that can be used to write router configuration generators.
peval, in a command loop,
inputs policy expressions,
expands the AS sets, route sets, filter sets, AS numbers 
(unless specified otherwise using command line options),
evaluates the resulting expression
and outputs the result.
Instead of a command loop,
the expression can be given as an argument.
.PP
By default peval evaluates the expression for ipv4 unicast address family. Also the list of address families can be specified in order to restrict the results. The syntax is: afi <afi_list> <mp-filter>.
.PP
<afi_list> is a comma separated list of address families. The supported ones are: ipv4, ipv4.unicast (same as ipv4), ipv4.multicast, ipv6, ipv6.unicast (same as ipv6), ipv6.multicast, any.unicast (ipv4.unicast, ipv6.unicast), any.multicast (ipv4.multicast, ipv6.multicast), any (ipv4.unicast, ipv4.multicast, ipv6.unicast, ipv6.multicast). If the results are to be restricted to uncomparable address families (ipv4, ipv6), in the output they will be joined using 'OR' clause.
.PP
<mp-filter> is a valid RPSLng filter over AS-numbers, ipv4/ipv6 prefixes, range operators, filter-sets, route-sets, as-sets and logical operators. For more details about mp-filter format see RPSLng documentation.
.PP
.SH IRRToolSet COMMON OPTIONS
.RS
.IP -help
Print summary of command-line options and abort.
.IP "\-T [whois_query | whois_response | input | all]"
Trace the argument. Useful for debugging.
.IP "\-D <integer>"
Print debugging messages of debug channel <integer> 
(if compiled with DEBUG defined).
.IP "\-version"
Print the version number and quit.
.IP "\-h <host-name>"
Establish a whois connection to host <host-name>.
The default is whois.radb.net.
.IP "\-p <port-no>"
Establish a whois connection to port <port-no>.
The default is 43.
.IP "\-protocol <protocol>"
Use the <protocol> to connect to the IRR server. <protocol> can be irrd(rawhoisd), ripe(bird), and ripe_perl. The default is irrd. The ripe_perl option should be used with the old ripe server.
.IP "\-s <source-list>"
Consider the sources specified in the comma separated <source-list>.
If an object is defined in multiple sources in <source-list>,
peval uses the definition first encountered in <source-list>
from left to right.
.IP "\-f <file-name>" 
IRR cache file. You can have any RPSL object in this file, except route
objects. 
They will override these objects in IRR.
This option is intended for private objects, or to test new public objects 
before publishing. You can specify more than one cache file by specifying this
option repeatedly.
.IP \-rusage
Print resource usage upon completion.
.IP "\-prompt <string>"
Set the interactive prompt to <string>.
.IP "\-ignore_errors"
Do not print error and warning messages due to communication to the
database server or parsing policy objects.
.IP "\-report_errors"
Print error and warning messages due to communication to the
database server or parsing policy objects.
.RE
.SH OTHER OPTIONS
.RS
.IP -no-as-set
Do not expand AS sets into AS numbers.
.IP -no-route-set
Do not expand route sets into prefixes.
.IP -no-as
Do not expand AS numbers into prefixes.
.IP -none
Do not expand anything.
.IP -symbolic
Do a prior symbolic evaluation, then do the expansions and then
re-evaluate.
This may be faster for some policies.
.IP -compressed
Print prefix lists using the more specific operators. 
Otherwise,
an expression like 128.9.0.0/16^24 will result in 256 prefixes being printed.
.IP "\-e <expression>"
Evaluate expression.
.RE
.SH ENVIRONMENT VARIABLES
.RS
.IP IRR_HOST
Specifies the IRR host to connect.
.IP IRR_PORT
Specifies the IRR port number to connect.
.IP IRR_SOURCES
Specifies the source list (comma separated) to consider.
.RE
.PP
Command line options take precedence over environment variables.
.SH EXAMPLES
.PP
Perform only symbolic evaluation:
.PP
.nf
% ./peval -none
peval> (AS1 OR AS2) AND AS1
((AS1))
peval> AS1 AND NOT AS1
NOT ANY
peval> 
.fi
.PP
Expand AS macros and evaluate:
.PP
.nf
% ./peval -no-as AS-ANSCT
((AS1326 AS2002 AS2538 AS2752 AS3723))
.fi
.PP
Expand all and evaluate 
(e.g. multihomed routes to AS1220 and AS226):
.PP
.nf
% ./peval 'AS1220 AND AS226'
({192.237.125.0/24, 192.92.56.0/24, 196.2.41.0/24, 196.2.40.0/24})
.fi
.PP 
Evaluate the filter for ipv4 and ipv6 address families:
.PP
.nf
% ./peval 'afi ipv6, ipv4 rs-rpslng'
({1001:0:0:0:0:0:0:0/35, 1:0:0:0:0:0:0:0/35})
 OR ({1.0.0.0/8, 1.2.3.0/24})
.SH ERROR AND WARNING MESSAGES
Error: Internal error.
.RS
.PP
Fails if couldn't evaluate the filter to resulting normal expression.
.PP
.RE
Unknown protocol!
.RS
.PP
Connection to IRR server failed. This can be caused by various reasons, please see Errors manpage, IRR Communication errors.
.PP
.RE
***Error: badly formed filter.
.RS
.PP
syntax error in RPSL filter specification. 
.PP
.RE
For more error descriptions, please see 
.B
Errors manual page.
.SH AUTHORS
Cengiz Alaettinoglu <cengiz@isi.edu>
.PP
Katie Petrusha <katie@ripe.net>


src/peval/peval.cc
//  $Id$
// Copyright (c) 2001,2002                        RIPE NCC
//
// All Rights Reserved
//
// Permission to use, copy, modify, and distribute this software and its
// documentation for any purpose and without fee is hereby granted,
// provided that the above copyright notice appear in all copies and that
// both that copyright notice and this permission notice appear in
// supporting documentation, and that the name of the author not be
// used in advertising or publicity pertaining to distribution of the
// software without specific, written prior permission.
//
// THE AUTHOR DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING
// ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS; IN NO EVENT SHALL
// AUTHOR BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY
// DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN
// AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
// OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
//
//
//  Copyright (c) 1994 by the University of Southern California
//  All rights reserved.
//
//    Permission is hereby granted, free of charge, to any person obtaining a copy
//    of this software and associated documentation files (the "Software"), to deal
//    in the Software without restriction, including without limitation the rights
//    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//    copies of the Software, and to permit persons to whom the Software is
//    furnished to do so, subject to the following conditions:
//
//    The above copyright notice and this permission notice shall be included in
//    all copies or substantial portions of the Software.
//
//    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
//    THE SOFTWARE.
//
//  Questions concerning this software should be directed to 
//  irrtoolset@cs.usc.edu.
//
//  Author(s): Cengiz Alaettinoglu <cengiz@ISI.EDU>

#include "config.h"
#include <cstdlib>
#include <cstring>

extern "C" {
#if HAVE_UNISTD_H
#   include <unistd.h>
#endif
}
#include <iostream>
#include <iomanip>

#include <cstdio>

#if HAVE_LIBREADLINE && HAVE_LIBHISTORY
extern "C" {
extern char *readline (char *);
extern void add_history (char *);
}
#endif // HAVE_LIBREADLINE and HAVE_LIBHISTORY

#include "irr/irr.hh"
#include "irr/rawhoisc.hh"
#include "irrutil/debug.hh"
#include "irrutil/trace.hh"
#include "irrutil/rusage.hh"
#include "irrutil/Argv.hh"
#include "irrutil/version.hh"
#include "rpsl/schema.hh"
#include "normalform/NE.hh"
#include "re2dfa/regexp_nf.hh"

using namespace std;

#define EXPAND_ASSets         0x000001
#define EXPAND_RSSets         0x000002
#define EXPAND_AS             0x000004
#define EXPAND_NOTHING        0
#define EXPAND_ALL            (~0)

bool opt_rusage                  = false;
Rusage ru(clog, &opt_rusage);

char *opt_prompt                 = (char *)"peval> ";
int  opt_expand                  = EXPAND_ALL;
int  opt_symbolic                = 0;
bool opt_asdot                   = false;

const int SIZE = 8*1024;
char base[SIZE] = "peval: ";
char temp[SIZE];
char safe_base[SIZE];
char *irrfilter;
char *cut;

void evaluate() {
   if (opt_expand & EXPAND_ASSets)
      regexp_nf::expandASSets();

   strcat(irrfilter, "\n\n");
   // Was: safe_base = base;
   memcpy(safe_base, base, SIZE);

   cut = strstr(irrfilter, "afi");
   if (cut && isspace(*(cut+3))) {
     strcat (temp, "mp-");
     strcat (temp, base);
     // Was: base = temp;
     memcpy(base, temp, SIZE);
     // Was: bzero(temp, SIZE);
     memset(temp, '\0', SIZE);
   }

   Object *o = new Object;
   o->scan(base, strlen(base));
   if (o->has_error) {
      delete o;
      // Was: base = safe_base;
      memcpy(base, safe_base, SIZE);
      return;
   }

   if (strcmp(o->type->getName(), "mp-peval") == 0) {

     AttrIterator<AttrMPPeval> itr(o, "mp-peval");

     NormalExpression *ne = 
        NormalExpression::evaluate(itr()->filter, ~0, opt_expand);

     if (ne)
        cout << *ne << endl;
     else
        cerr << "Error: Internal error." << endl;
	
     delete ne;
     delete o;

   } else {
 
   AttrIterator<AttrFilter> itr(o, "peval");

   NormalExpression *ne = 
      NormalExpression::evaluate(itr()->filter, ~0, opt_expand);

   if (ne)
      cout << *ne << endl;
   else
      cerr << "Error: Internal error." << endl;
  
   delete ne;
   delete o;

   }
   // Was: base = safe_base;
   memcpy(base, safe_base, SIZE);
}

int start_tracing(char *dst, char *key, char *nextArg) {
   if (nextArg) {
      trace.enable(nextArg);
      return 1; // return 1 to signify nextArg is used by us
   }
   return 0; 
}

int start_debugging(char *dst, char *key, char *nextArg) {
   if (nextArg) {
      Debug(dbg.enable(atoi(nextArg)));
      return 1; // return 1 to signify nextArg is used by us
   }
   return 0;
}

int skipASSets(char *dst, char *key, char *nextArg) {
   opt_expand &= ~EXPAND_ASSets;
   return 0;
}

int skipRSSets(char *dst, char *key, char *nextArg) {
   opt_expand &= ~EXPAND_RSSets;
   return 0;
}

int skipAS(char *dst, char *key, char *nextArg) {
   opt_expand &= ~EXPAND_AS;
   return 0;
}

int skipAll(char *dst, char *key, char *nextArg) {
   opt_expand = EXPAND_NOTHING;
   return 0;
}

int expression(char *dst, char *key, char *nextArg) {
   if (!nextArg) return 0;

   strcpy(irrfilter, nextArg);
   return 1;
}

void init_and_set_options (int argc, char **argv, char **envp) {
   ArgvInfo argTable[] = {
      // RAToolSet common arguments
      // key, type, src, dst, help
      {"-T", ARGV_FUNC, (char *) &start_tracing,      (char *) NULL, 
       "Start tracing the next argument"},
      {"-D", ARGV_FUNC, (char *) &start_debugging,    (char *) NULL, 
       "Start debugging the next argument"},
      {"-version", ARGV_FUNC, (char *) &version,      (char *) NULL,
       "Show version"},
      {"-rusage", ARGV_BOOL, (char *) NULL,           (char *) &opt_rusage,
       "On termination print resource usage"},
      {"-prompt", ARGV_STRING,  (char *) NULL,        (char *) &opt_prompt,
       "Prompt"},

      IRR_COMMAND_LINE_OPTIONS,

      {"-asdot", ARGV_BOOL, (char *) NULL, (char *) &opt_asdot,
       "print AS numbers in asdot format."},

      // peval specific arguments
      {"-symbolic",  ARGV_CONSTANT, (char *)1, (char *)&opt_symbolic, 
       "Symbolic"},
      {"-no-as-set", ARGV_FUNC, (char *)skipASSets, (char *)NULL,
       "Do not expand AS sets"},
      {"-no-route-set", ARGV_FUNC, (char *)skipRSSets, (char *)NULL, 
       "Do not expand route sets"},
      {"-no-as", ARGV_FUNC, (char *)skipAS, (char *)NULL,
       "Do not expand AS numbers"},
      {"-none", ARGV_FUNC, (char *)skipAll, (char *)NULL,
       "Do not expand anything"},
      {"-compressed", ARGV_BOOL, 
       (char *) NULL, (char *) &RadixSet::compressedPrint,
       "Print prefix lists using the more specific operators."},
      {"-e", ARGV_FUNC, (char *)expression, (char *)NULL,
       "Expression"},
      {(char *) NULL, ARGV_END, (char *) NULL, (char *) NULL, (char *) NULL}
   };

   IRR::handleEnvironmentVariables(envp);

   if (ParseArgv(&argc, argv, argTable, ARGV_NO_ABBREV) != ARGV_OK) {
      cerr << endl;
      exit(1);
   }

   if (! (irr = IRR::newClient())) {
      cerr << "Unknown protocol!" << endl;
      exit(1);
   }

   // if there are remaining arguments
   // the first one is the expression
   switch (argc) {
   case 2:
      strcpy(irrfilter, argv[1]);
      break;
   case 1:
      break;
   default:
      cerr << "Wrong number of arguments..." << endl;
      exit(-1);
   }

   if (*irrfilter) {
      evaluate();
      exit(0);
   }

   // have a prompt only if the input is coming from a tty
   if (!isatty(fileno(stdin)) || !isatty(fileno(stdout)))
      opt_prompt = (char *) NULL;
}

#if HAVE_LIBREADLINE && HAVE_LIBHISTORY
/* A static variable for holding the line. */
static char *line_read = (char *)NULL;

/* Read a string, and return a pointer to it.  Returns NULL on EOF. */
char *
rl_gets (char *prompt)
{
  /* If the buffer has already been allocated, return the memory
     to the free pool. */
  if (line_read)
    {
      free (line_read);
      line_read = (char *)NULL;
    }

  /* Get a line from the user. */
  line_read = readline (prompt);

  /* If the line has any text in it, save it on the history. */
  if (line_read && *line_read)
    add_history (line_read);

  return (line_read);
}
#endif // HAVE_READLINE

int main (int argc, char **argv, char **envp) {
   int newSize = strlen(base);
   irrfilter = base + newSize;
   newSize = SIZE - newSize - 3;

   schema.initialize();
   init_and_set_options(argc, argv, envp);

   while (1) {
#if HAVE_LIBREADLINE && HAVE_LIBHISTORY
      char *line = rl_gets(opt_prompt);
      if (!line)
	 break;
      strncpy(irrfilter, line, newSize);
      evaluate();
#else // HAVE_READLINE
      if (opt_prompt)
	 cout << opt_prompt;
      if (cin.getline(irrfilter, newSize))
	 evaluate();
      else
	 break;
#endif
   }

   cout << endl;

   CLASS_DEBUG_MEMORY_PRINT(Set);
   CLASS_DEBUG_MEMORY_PRINT(NormalTerm);
   CLASS_DEBUG_MEMORY_PRINT(NormalExpression);
   CLASS_DEBUG_MEMORY_PRINT(SymbolConjunct);
}
